formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/ProtoBuf.kt

```diff
@@ -281,8 +285,8 @@ public class ProtoBuf(
         @Suppress("UNCHECKED_CAST")
         override fun <T> decodeSerializableValue(deserializer: DeserializationStrategy<T>): T = when {
             // encode maps as collection of map entries, not merged collection of key-values
-            deserializer is MapLikeSerializer<*, *, *, *> -> {
-                val serializer = (deserializer as MapLikeSerializer<Any?, Any?, T, *>)
+            deserializer is MapLikeSerializer<*, *, *> -> {
```

Comment:
Hm, it actually seems that approach with `.toMap` is faulty: if one have custom serializer with `StructureKind.MAP`, this method will still return `Map` instance — i.e. in your example with StringPairSerializer, we would be able to serialize `StringPair` with `toMap` function, **but** in deserialize we get an instance of `Map` instead of `StringPair`.